### ISSUE TITLE
IRF Background interpolation produces peaks

### ISSUE THREAD

**Issue by StFroese** [gammapy/gammapy#5085]
Hi all 😄 

So, I'm currently trying to create a dataset from an observation. I have full enclosure IRFs with a Background3DModel which is constructed from ON data using the exclusion background method (cutting out the source and using the wobble between obs to fill the background).

Somewhere down the line when you're using the `MapDatasetMaker` the function background model will be evaluated for the energy edges of the geom and the coords of the geom, e.g. `Background3D.evaluate` will be called.

https://github.com/gammapy/gammapy/blob/90459a5bb99d7b22f0230627c05268d94212ffbc/gammapy/irf/core.py#L309-L329

This forces `linear` interpolation between energy bins. Which comes with two problem I think:
1. Does it make sense to interpolate a quantity like bkg rate *linear* if the grid point of the interpolation are given on a log scale (energy)?

2. If you construct a BackgroundModel with the exclusion method and you have limited observation time in your data, you will have a sparse background in some energy bins simply because of the powerlaw. See here

![bkg_from_exclusion](https://github.com/gammapy/gammapy/assets/44728015/a1436581-fa3a-4bec-98ed-e92ea5b7db90)

If you interpolate this linear you get peeks in your background because you evaluate your model a the energy axis edges

![bkg_interpoaltion_linear.pdf](https://github.com/gammapy/gammapy/files/14195447/bkg_interpoaltion_linear.pdf)

(caution: there are not just the peek values in the maps, the peek value is just so high that the color of the smaller values is almost the same as for 0.0)

# Solution
1. interpolate in log?

2. Let the `method` kwarg for the interpolation be an kwarg for the  call of the `interpolate_loglog` method
```python
def integrate_log_log(self, axis_name, method='linear', **kwargs): 
```
This produces the following background when choosing `method='nearest'`, which is still sparse but has no high peeks overshadowing the rest of the model (the peeks you see differ multiple orders of magnitude from the "`linear`" ones)
![bkg_interpolation_nearest.pdf](https://github.com/gammapy/gammapy/files/14195479/bkg_interpolation_nearest.pdf)

**Comment by StFroese** [comment]
@adonath can you take a look? 😊   seems like you wrote this code 4 years ago hahaha

**Comment by adonath** [comment]
Thanks @StFroese, I'm pretty sure the code used to do log-log interpolation. The log energy bins is still handled here: https://github.com/gammapy/gammapy/blob/90459a5bb99d7b22f0230627c05268d94212ffbc/gammapy/irf/core.py#L197

But it seems the log scaling of the background is missing. Usually it is declared in the `default_interp_kwargs` (https://github.com/gammapy/gammapy/blob/90459a5bb99d7b22f0230627c05268d94212ffbc/gammapy/irf/core.py#L60), which does not seem to be the case. The "linear" mode is a bit misleading, because it actually means linear interpolation in log-log space.

I'll double check again to see what is happening in detail.

**Comment by StFroese** [comment]
@adonath thanks for taking a look!
Maybe I missed something about the log interpolation, it's really misleading with the method being linear but there's still something weird about it I think. 
Thanks in advance for double checking:)

**Comment by adonath** [comment]
@StFroese the situation is as follows: non linear interpolation is Gammay is handled by transforming the data as well as coordinate grids:

The coordinate axes are transformed here:
https://github.com/gammapy/gammapy/blob/90459a5bb99d7b22f0230627c05268d94212ffbc/gammapy/irf/core.py#L197

The transformation "scale" for the data is defined here:
https://github.com/gammapy/gammapy/blob/90459a5bb99d7b22f0230627c05268d94212ffbc/gammapy/irf/background.py#L35

So if you use `interp="log"` for the energy axis as well as the default interpolation kwargs, it should do log log interpolation. I'm not sure where the difference comes from in the image you show. Maybe it comes from the handling of small values in the log scaling? See e.g. https://github.com/gammapy/gammapy/blob/main/gammapy/utils/interpolation.py#L189

**Comment by adonath** [comment]
@StFroese Just looking at images, this is a very tricky case. There is clearly not enough statistics to provide any stable background estimate. Typically this is handled with some pre-process such as smoothing or adaptive smoothing. You could also try to interpolate across the energy axis. But the shape of the background for IACTs at high energies is complex. See for example https://www.aanda.org/articles/aa/abs/2019/12/aa36452-19/aa36452-19.html (especially https://www.aanda.org/articles/aa/full_html/2019/12/aa36452-19/F12.html). I think the production of background models from sparse data is mostly an un-solved problem. Most people just fall back to an on-off analysis in this case.

**Comment by StFroese** [comment]
@adonath I agree, it's really difficult to generate background for low statistics. I can try smooth it but I also think that interpolation between energies won't work because of the donut shape at high energies. Another way would maybe to take a look at the exposure maps and fold them with the psf and energy dispersion to get a rough estimate and fit a model based on this to the sparse maps

**Comment by StFroese** [comment]
@adonath I'm still a bit confused. Isn't this line here exactly allowing interpolation between energies, which should not be allowed by default?
https://github.com/gammapy/gammapy/blob/90459a5bb99d7b22f0230627c05268d94212ffbc/gammapy/irf/core.py#L327

I also did a little bit of testing and the issue seems to not come from low statistics. 
I generated a background model which has enough statistics by looking at up to 100GeV only:

![bkg_model_lowE.pdf](https://github.com/gammapy/gammapy/files/14376571/bkg_model_lowE.pdf)

(don't get confused by the axes here, it says RaDec but it's actually AltAz)

And when I apply this model to a dataset, I still get the hot pixel in the highest energy bin:

![bkg_applied.pdf](https://github.com/gammapy/gammapy/files/14376583/bkg_applied.pdf)

**Comment by adonath** [comment]
> I'm still a bit confused. Isn't this line here exactly allowing interpolation between energies, which should not be allowed by default?

Yes, this is the intended behavior. It should be allowed by default. The point was just that by default the interpolation happens linearly in log-log, which is a reasonable behavior for power-law shaped functions in energy.

> I generated a background model which has enough statistics by looking at up to 100GeV only:

At high offsets you will always encounter low statistics. I have typically seen hot pixels when the data is extrapolated beyond the valid energy range. So this might be something to double check. 

Gammapy assumes that IRFs are noise free and reasonably smooth. So the responsibility is on the person creating the ORF to create a "well behaved" IRF, because it requires a lot of domain knowledge to come up with a good and stable background model. So you have to solve this problem using pre-processing of the background model. For example by spatially smoothing per energy bin or smoothing along the energy axis using some local power-law assumption. 

From the Gammapy side we could maybe move the hard-coded `method="linear"` and make it configurable via the `default_interp_kwargs`.

**Comment by StFroese** [comment]
@adonath thanks a lot for the explanations! :)

**Comment by AtreyeeS** [comment]
What is the remaining action here @adonath @StFroese ?

**Comment by StFroese** [comment]
@AtreyeeS I think best is to allow for nearest interpolation if wanted by the user :)
